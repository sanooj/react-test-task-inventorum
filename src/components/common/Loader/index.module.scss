@import "./../../../styles/variables";

.loader {
	border: 5px solid $white;
	border-radius: 50%;
	border-top: 5px solid $gun-powder-blue;
	width: 30px;
	height: 30px;
	animation: spin 2s linear infinite;
	margin: auto;
	position: relative;
	left: 0;
	right: 0;
	top: 44%;
	bottom: 0;

	&--inline {
		margin: 30px auto;
	}

	&-wrapper {
		height: 100%;
		z-index: 1500;
		position: fixed;
		width: 100%;
		top: 0;
		left: 0;
		background-color: rgba($black, 0.7);
	}

	.loading-text {
		position: absolute;
		top: 50%;
		height: 100%;
		width: 100%;
		text-align: center;
		color: $white;
	}
}

@keyframes spin {
	0% {
		transform: rotate(0deg);
	}
  
	100% {
		transform: rotate(360deg);
	}
}
