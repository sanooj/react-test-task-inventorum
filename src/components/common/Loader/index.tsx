import styles from "./index.module.scss";

/**
 * Loader
 * @param type : loader type 
 * @returns : JSX element
 */
const Loader = ({ type }: { type?: "inline" }): JSX.Element => {
	return (
		<>
			{type === "inline" ? (
				<div
					className={`${styles["loader"]} ${styles["loader--inline"]}`}
					data-testid="inlineLoader"
				></div>
			) : (
				<div className={styles["loader-wrapper"]} data-testid="fullPageLoader">
					<div className={styles["loader"]}></div>
					<div className={styles["loading-text"]}>
						Please wait.
						<br />
						Loading ....
					</div>
				</div>
			)}
		</>
	);
};

export default Loader;
