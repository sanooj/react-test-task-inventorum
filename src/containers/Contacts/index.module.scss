@import "./../../styles/variables";
@import "./../../styles/mixins";
@import "./../../styles/typo";

.main {
	display: grid;
	grid-template-columns: 181px auto;
	min-height: 100vh;
}

.form-field {
	display: flex;

	&__item {
		&--search {
			margin-right: 13px;
			flex: 1;
			margin-bottom: 10px;
			background-color: #ebecee;
			padding: 0 12px;
			border-radius: 3px;

			input {
				background-color: transparent;
				border: 0;
				height: 35px;
				width: 100%;
				@include font-size(13);
			}
		}
	}
}
