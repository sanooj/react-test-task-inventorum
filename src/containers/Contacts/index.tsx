import React, { ChangeEvent, useEffect, useMemo, useState } from "react";
import { SearchOutlined } from "@ant-design/icons";
import { Input } from "antd";

import styles from "./index.module.scss";
import DefaultLayout from "layout/DefaultLayout";
import Users from "components/custom/contacts/Users";
import ButtonTheme from "components/common/ButtonTheme";
import { services } from "services/Services";
import { UserInterface, UsersDataInterface } from "@interfaces/interfaces";
import debouce from "lodash.debounce";
import { withAlertHOC } from "components/common/withAlertHOC";

/**
 * Contact page
 * @param showMessage : HOC Alert Component 
 * @returns JSX Element
 */
const Contacts = ({ showMessage }: { showMessage: any }): JSX.Element => {
	// users state
	const [usersData, setUsersData] = useState<UserInterface[]>([]);
	const [usersItem, setUsersItem] = useState<UserInterface[]>([]);

	/**
	 * get User Data
	 */
	const getUserData = (params: string) => {
		services
			.getUserData(`users?${params}`)
			.then(({ data }: UsersDataInterface) => {
				setUsersData(data);
				setUsersItem(data);
			}).catch((err) => {
        	showMessage(err, "error");
      });
	};

	// initial loading
	useEffect(() => {
    // loading complete user data
		getUserData("per_page=12");
	}, []);

	/**
	 * On Search method
	 * @param e : change event
	 */
	const onSearch = (e: ChangeEvent<HTMLInputElement>) => {
		// search value
		const targetValue = e.target.value;

		// filter search value
		const searchedUserResults = usersData.filter(
			({ first_name, last_name }) => {
				const fullName = `${first_name} ${last_name}`;
				return fullName?.toLowerCase().includes(targetValue?.toLowerCase());
			}
		);
		setUsersItem(searchedUserResults);
	};

	/**
	 * Debounce ( Memo on users data loaded)
	 */
	const debouncedSearchResults = useMemo(() => {
		return debouce(onSearch, 300);
	}, [usersData]);

	/**
	 * useEffect clean up any side effects from debounce when our component gets unmounted
	 */
	useEffect(() => {
		return () => {
			debouncedSearchResults.cancel();
		};
	});

	return (
		<DefaultLayout heading="Users">
			<section>
				<div className={styles["form-field"]} data-testid="searchField">
					<div
						className={`${styles["form-field__item"]} ${styles["form-field__item--search"]}`}
					>
						<Input
							type={"search"}
							placeholder="Search user"
							prefix={<SearchOutlined />}
							onChange={debouncedSearchResults}
							prefixCls="search"
              data-testid="searchInput"
						/>
					</div>
					<div
						className={`${styles["form-field__item"]} ${styles["form-field__item--add-button"]}`}
					>
						<ButtonTheme>Add</ButtonTheme>
					</div>
				</div>

				<Users userData={usersItem} />
			</section>
		</DefaultLayout>
	);
};

export default withAlertHOC(Contacts);
