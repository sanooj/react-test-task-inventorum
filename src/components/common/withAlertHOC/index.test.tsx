import React, { Suspense } from "react";
import { render, screen } from "@testing-library/react";
import { withAlertHOC } from ".";
import Contacts from "containers/Contacts";
import { BrowserRouter } from "react-router-dom";
import Loader from "../Loader";


describe("Alert HOC", () => {

   const MockComponent = withAlertHOC(() => (
			<BrowserRouter>
				<Suspense fallback={<Loader />}>
					<Contacts />
				</Suspense>
			</BrowserRouter>
		));
   
	it("renders snapshot", () => {
		const { container } = render(<MockComponent />);
		expect(container).toMatchSnapshot();
	});

});
