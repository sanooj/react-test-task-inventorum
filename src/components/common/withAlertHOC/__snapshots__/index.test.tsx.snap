// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert HOC renders snapshot 1`] = `
<div>
  <div
    class="main"
    role="main"
  >
    <aside
      class="side-navigation"
    >
      <a
        class="logo"
        href="/"
      >
        <img
          alt="Inventorum"
          data-testid="logo"
          src="/images/logo.svg"
        />
      </a>
      <nav
        class="nav"
        data-testid="navigation"
      >
        <a
          class="nav__item"
          href="/ponts-of-sale"
        >
          <i
            class="icon-monitor"
          />
           
          Point of sales
        </a>
        <a
          class="nav__item"
          href="/invoice-returns"
        >
          <i
            class="icon-monitor"
          />
           
          Invoices & returns
        </a>
        <a
          class="nav__item"
          href="/contacts"
        >
          <i
            class="icon-monitor"
          />
           
          Contacts
        </a>
        <a
          class="nav__item"
          href="/inventory"
        >
          <i
            class="icon-monitor"
          />
           
          Inventory
        </a>
        <a
          class="nav__item"
          href="/reports"
        >
          <i
            class="icon-monitor"
          />
           
          Reports
        </a>
        <a
          class="nav__item"
          href="/settings"
        >
          <i
            class="icon-monitor"
          />
           
          Settings
        </a>
        <a
          class="nav__item"
          href="/imprint"
        >
          <i
            class="icon-monitor"
          />
           
          Imprint
        </a>
      </nav>
    </aside>
    <section
      class="main__section"
      data-testid="appSection"
    >
      <header
        class="header"
      >
        <div
          class="header__panel header__panel--left"
        >
          <i
            class="icon-burger-menu"
          />
          <h2
            data-testid="mainHeading"
          >
            Users
          </h2>
        </div>
        <div
          class="header__panel header__panel--right"
        >
          <i
            class="user-icon"
          >
            <img
              alt="profile"
              src="/images/profile-lock.svg"
            />
             
          </i>
          <a>
            Patricia Kasse 
            <i
              class="icon-triangle"
            />
          </a>
        </div>
      </header>
      <div
        class="container"
      >
        <section>
          <div
            class="form-field"
            data-testid="searchField"
          >
            <div
              class="form-field__item form-field__item--search"
            >
              <span
                class="search-affix-wrapper"
              >
                <span
                  class="search-prefix"
                >
                  <span
                    aria-label="search"
                    class="anticon anticon-search"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="search"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                      />
                    </svg>
                  </span>
                </span>
                <input
                  class="search"
                  data-testid="searchInput"
                  placeholder="Search user"
                  type="search"
                  value=""
                />
              </span>
            </div>
            <div
              class="form-field__item form-field__item--add-button"
            >
              <button
                class="ant-btn ant-btn-round ant-btn-primary ant-btn-lg primary-theme theme-button--primary"
                data-testid="button"
                type="button"
              >
                <span>
                  Add
                </span>
              </button>
            </div>
          </div>
          <div
            class="users-panel"
          >
            <div
              class="user-wrapper"
              data-testid="userWraper"
            >
              <div
                class="loader loader--inline"
                data-testid="inlineLoader"
              />
            </div>
          </div>
        </section>
      </div>
    </section>
  </div>
</div>
`;
